src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataReaderExtensions.cs

```diff
@@ -97,22 +105,14 @@ public static ODataItemBase ReadResourceOrResourceSet(this ODataReader reader)
                         Contract.Assert(resourceSet != null, "ResourceSet should never be null.");
 
                         ODataResourceSetWrapper resourceSetWrapper = new ODataResourceSetWrapper(resourceSet);
-                        if (itemsStack.Count > 0)
-                        {
-                            ODataNestedResourceInfoWrapper parentNestedResourceInfo = (ODataNestedResourceInfoWrapper)itemsStack.Peek();
-                            Contract.Assert(parentNestedResourceInfo != null, "this has to be an inner resource set. inner resource sets always have a nested resource info.");
-                            Contract.Assert(parentNestedResourceInfo.NestedResourceInfo.IsCollection == true, "Only collection nested properties can contain resource set as their child.");
-                            parentNestedResourceInfo.NestedItems.Add(resourceSetWrapper);
-                        }
-                        else
-                        {
-                            topLevelItem = resourceSetWrapper;
-                        }
+                        
+                        AddNestedResourceInfo(ref topLevelItem, itemsStack, resourceSetWrapper);
```

Comment:
>AddNestedResourceInfo [](start = 24, length = 21)

Consider not passing topLevelItem to AddNestedResourceInfo, and calling (here and in DeltaresourceSetStart) as:

if (itemsStack.Count > 0)
{
    AddNestedResourceInfo(itemsStack, resourceSetWrapper);
}
else
{
    topLevelItem = resourceSetWrapper;
}

It's a judgement call but, while i like having methods that encapsulate common code, in this case having AddNestedResourceInfo maybe add the nested resource info otherwise set the topLevelItem is confusing the intent of that method.  It works, but over time may make the code harder to read/maintain.